### ISSUE TITLE
cargo-audit fails to check yanked status behind corporate transparent proxy

### ISSUE THREAD

**Issue by arvid-norlander** [rustsec/rustsec#988]
We have a transparent (no need to set environment variables) HTTP/HTTPS proxy at work. The certificates are installed system wide in `/etc/ca-certificates`.

cargo-audit does not seem to be able to properly check yanking status (and it doesn't give enough error output to properly diagnose the issue):

```console
$ cargo audit bin ~/.cargo/bin/cargo-audit
    Fetching advisory database from `https://github.com/RustSec/advisory-db.git`
      Loaded 563 security advisories (from /home/tmgarn/.cargo/advisory-db)
    Updating crates.io index
       Found 'cargo auditable' data in /home/tmgarn/.cargo/bin/cargo-audit (260 dependencies)
error: couldn't check if the package is yanked: registry: Failed to retrieve abscissa_core from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve abscissa_derive from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve addr2line from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve adler from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve ahash from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve aho-corasick from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve arc-swap from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve arrayvec from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve async-compression from crates.io index: registry: request could not be completed in the allotted timeframe
[...]
error: couldn't check if the package is yanked: registry: Failed to retrieve wait-timeout from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve walkdir from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve want from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve winnow from crates.io index: registry: request could not be completed in the allotted timeframe
error: couldn't check if the package is yanked: registry: Failed to retrieve zeroize from crates.io index: registry: request could not be completed in the allotted timeframe
```

This affects both normal (workspace) and `bin` operations. It is possible something is wrong with the proxy, but I can't tell unless it gives me some more details in the error message (was it a TCP timeout? did it start getting the file and not finish? did it fail to connect? how long did it wait? What URL did it try to access when this happened?)

As such one of these is true:
* There is an incompatibility with transparent HTTPS proxies in cargo-audit that needs to be fixed.
* There is an issue with the proxy, and cargo-audit doesn't report detailed enough error information to properly diagnose it.

Versions:

* cargo-audit 0.18.1 (installed via cargo install)
* rust 1.72.0 (installed via rustup)
* Ubuntu 22.04

**Comment by Shnatsel** [comment]
Thank you for the report!

As a quick fix, `CARGO_REGISTRIES_CRATES_IO_PROTOCOL=git cargo audit` should work, but I'd like to get to the bottom of this. Sadly there is not much `cargo audit` can log about the situation - all it knows is that it sends the request and never receives a response. 

`cargo-audit` does read the system-wide certificates. So my guess is that your proxy doesn't support HTTP/2. Does removing this line help? 

https://github.com/rustsec/rustsec/blob/ed9c2224f95201cd89ddf76ae9a57cce79181948/rustsec/src/cached_index.rs#L94

`cargo audit` must always get `audit` as the first argument, so you'll need to run `cargo run --release -- audit` to test the change.

**Comment by arvid-norlander** [comment]
> As a quick fix, CARGO_REGISTRIES_CRATES_IO_PROTOCOL=git cargo audit should work, 

It does, however cargo itself works fine with sparse index for me.

> Sadly there is not much cargo audit can log about the situation - all it knows is that it sends the request and never receives a response.

Useful information to display would be something like:
```
error: couldn't check if the package is yanked: registry: Failed to retrieve abscissa_core from crates.io index: registry: request could not be completed in the allotted timeframe)
  tried to download https://example.com/...
  got OS error code 1234 (timeout)
```

Clearly there is some useful information that *could* be displayed that isn't. It might require changes to dependencies.

> cargo-audit does read the system-wide certificates. So my guess is that your proxy doesn't support HTTP/2. Does removing this line help?

I just get tens of thousands of lines of:

```
2023-09-06T07:47:51.540305Z  WARN hyper::client::client: Connection is HTTP/1, but request requires HTTP/2
```

I tried to redirect to a log file, but I aborted it after it reached multiple GB.

**Comment by Shnatsel** [comment]
Ah, that was not sufficient to change from HTTP/2 to HTTP 1.1 because `tame-index` also forces HTTP/2, here: https://github.com/EmbarkStudios/tame-index/blob/f10814bfc5bb005d1d471edbc24bc96388319c52/src/index/sparse.rs#L100

Let me put together a branch with all the HTTP/2 bits removed and I'll let you test it. I'm 80% certain your proxy just doesn't support HTTP/2 and that's what causes the issue.

> Clearly there is some useful information that could be displayed that isn't. It might require changes to dependencies.

Fair enough. We can probably get debug output from `reqwest` using `log` or `tracing` without too much trouble.

**Comment by Shnatsel** [comment]
Please try this branch: https://github.com/rustsec/rustsec/tree/http1

It uses a fork of `tame-index` to drop all requirements of HTTP/2.

**Comment by arvid-norlander** [comment]
> Please try this branch: https://github.com/rustsec/rustsec/tree/http1
> 
> It uses a fork of `tame-index` to drop all requirements of HTTP/2.

On this branch `cargo run --bin cargo-audit --release -- audit` works perfectly (cargo complained if I didn't specify which binary to run).

So yes, it seems hoping that the world is HTTP/2 is too early. And I rate my chances of getting IT to fix our proxy to be approximately 0.

**Comment by Shnatsel** [comment]
It seems 8 years was not enough for proxies to start supporting it. Oh well.

I've opened https://github.com/EmbarkStudios/tame-index/pull/26 for `tame-index`, I'll also make a PR for `cargo audit` and release a new version.

**Comment by Shnatsel** [comment]
The fix has shipped in `rustsec` 0.28.1. No changes to `cargo-audit` itself were required, so I didn't bump its version. You can get the fix with `cargo install --force cargo-audit`.
